Website home (#92)
AdRoll/baker: content/en/docs/Performance/_index.md

---
title: "Performance"
weight: 150
date: 2020-11-20
description: >
  Baker performance test cases
---

### Read from S3 and write to local disk

> On a `c5.2xlarge` instance, Baker managed to read zstandard records from S3, uncompress them and

Reviewer: Not sure c5.2xlarge speaks to everybody, maybe we could add in parenthesis (or as a footer note as you prefer) that this is an AWS EC2 instance based on Intel Xeon Platinum 8000. Or maybe just that is an AWS instance type.
```suggestion
On a `c5.2xlarge` instance, Baker can read zstandard records from S3, uncompress them and
```

---
title: "Performance"
weight: 150
date: 2020-11-20
description: >
  Baker performance test cases
---

We report in this page some practical examples of Baker performances we could measure in the
NextRoll production environment.

Within NextRoll, Baker is often executed on AWS EC2 instances, and thus you find in this page
many references to
[EC2 instance types](https://aws.amazon.com/ec2/instance-types/) (`c5.2xlarge`, `c5.2xlarge`, etc.).

### Read from S3 and write to local disk

On an AWS EC2 instance of size `c5.2xlarge`, Baker can read zstandard records from S3, uncompress
them and apply a basic filtering logic, compressing them back on local files with zstandard at

> compression level 3 and long range mode at 27 using ~90% of capacity of each vCPU (8 in total) and

Reviewer: I wouldn't write long range mode without explaining what that means here since it's very technical and very specific.
But I'm not sure we should write that, nor compression level 3 for that matter.

Author: removed

~3.5GB of RAM.  

> It reads and writes a total of 94 million records in 8'51" (~178k r/w records per second).

Reviewer: Since we already wrote that baker reads and writes, we should probably just stick to `178k records per second`.
Again, we should round things up a bit in order to make it easily readable IMHO 
```suggestion
It reads and writes a total of 94 million records in less than 9 minutes, that's 178k records per second.
```

> On a `c5.2xlarge` instance (48vCPUs) the same test takes 2'2" (~775k r/w records per second).

Reviewer: Same here, the `'` `''` notation doesn't make for an easy read IMO
```suggestion
On a `c5.2xlarge` instance (48 vCPUs) the same test takes 2 minutes, so that's a speed of 775k records per second.
```

> For this test we use 711 zstd compressed files for a total of 17 GB of compressed size and 374 GB
> of uncompressed size. The average size of each record is ~4.5 KB.

Reviewer: Since we already say it's an average then the `~` symbol can be removed
```suggestion
For this test we use 711 zstd compressed files for a total of 17 GB of compressed size and 374 GB
of uncompressed size. The average size of each record is 4.5 KB.
```

### Read from S3 and write to DynamoDB (in the same region)

On a `c5.4xlarge` instance, Baker reads zstd compressed files from S3 writing to DynamoDB (configured
with 20k write capacity units) at an average speed of 60k records/s (the average size of each record

> is 4.3 KB) using less than 1 GB of memory and ~300% of the total CPU capacity (less than 20% for

Reviewer: This suggestion just removes `~`
```suggestion
is 4.3 KB) using less than 1 GB of memory and 300% of the total CPU capacity (less than 20% for
```

Not sure what _less than 20%_ means? _less than 20% of CPU usage_ maybe?

each core).

> The bottleneck here is the DynamoDB write capacity, so Baker can easily cope with an increased load
> just increasing the write capacity units in DynamoDB (up to 400k).

Reviewer: repetition of increased.
repetition of dynamodb
```suggestion
The bottleneck here is the DynamoDB write capacity, so Baker could handle the additional load
caused by increasing the write capacity, up to 400k.
```
I dont understand where does that 400k come from? if at 20k write capacity we use 20% of CPU, that means that we would be maxed out at 100% CPU if we multiply the write capacity by 5, so 100k write capacity.
What am i missing?

Author: ok, the number is the max write capacity you can set in dynamodb, not considering our cpu. We can clarify with `just increasing the write capacity units in DynamoDB (whose max value can get up to 400k).` or similar

Reviewer: Oh ok so I'm not sure It's super relevant for this page then. We can just say baker would be able to handle an additional load, without writing that number

Author: ```suggestion
The bottleneck here is the DynamoDB write capacity, so Baker could handle the additional load caused by
a possible increase in the write capacity.
```

### Read from Kinesis and write to DynamoDB (in the same region)

On a `c5.4xlarge` instance, we performed a test reading from a Kinesis stream with 130 shards and
writing to a DynamoDB table with 20k write capacity units. Baker is able to read and write more

> than 10k records per second (the avg size of each record is 4.5 KB) using less than 1 GB of RAM and
> ~400% of the total CPU capacity (less than 25% for each core).

Reviewer: ```suggestion
than 10k records per second (the average size of each record is 4.5 KB) using less than 1 GB of RAM and
around 400% of the total CPU capacity (each core being used at less than 25%).
```
